src/robot_new.cpp

```diff
@@ -0,0 +1,396 @@
+}
+
+void Robot::updateSubscribers(){
+    if (external_robot_state) {
+        ROS_WARN("Subscribing to external joint state");
+        joint_state_sub = nh->subscribe(topic_root+"/sensing/external_joint_states", 1, &Robot::JointState, this);
+    }
+    joint_target_sub = nh->subscribe(topic_root+"control/joint_targets", 100, &Robot::JointTarget, this);
+    controller_type_sub = nh->subscribe("/controller_type", 100, &Robot::controllerType, this);
+//     joint_state_sub = nh->subscribe("/joint_states", 100, &Robot::JointState, this);
+//    floating_base_sub = nh->subscribe(topic_root+"control/floating_base", 100, &Robot::FloatingBase, this);
+//    ik_srv = nh->advertiseService(topic_root+"control/ik", &Robot::InverseKinematicsService, this);
+//    ik_two_frames_srv = nh->advertiseService(topic_root+"control/ik_multiple_frames", &Robot::InverseKinematicsMultipleFramesService, this);
+//    fk_srv = nh->advertiseService(topic_root+"control/fk", &Robot::ForwardKinematicsService, this);
+//    freeze_srv = nh->advertiseService(topic_root+"control/freeze", &Robot::FreezeService, this);
+//    interactive_marker_sub = nh->subscribe(topic_root+"control/interactive_markers/feedback",1,&Robot::InteractiveMarkerFeedback, this);
+//    zero_joints_sub = nh->subscribe(topic_root+"control/zero_joints", 1, &Robot::ZeroJoints,this);
+}
+
+void Robot::init(string urdf_file_path, string viapoints_file_path, vector<string> joint_names_ordered) {
+
+    updatePublishers();
+    fmt = Eigen::IOFormat(4, 0, " ", ";\n", "", "", "[", "]");
+    nh->getParam("external_robot_target", external_robot_target);
+    nh->getParam("external_robot_state", external_robot_state);
+    nh->setParam("vr_puppet",false);
+
+    kinematics.init(urdf_file_path, viapoints_file_path, joint_names_ordered);
+
+    q.resize(kinematics.number_of_dofs);
+    qd.resize(kinematics.number_of_dofs);
+    qdd.resize(kinematics.number_of_dofs);
+
+    q_target.resize(kinematics.number_of_dofs);
+    qd_target.resize(kinematics.number_of_dofs);
+    qdd_target.resize(kinematics.number_of_dofs);
+
+    q_target_prev.resize(kinematics.number_of_dofs);
+    qd_target_prev.resize(kinematics.number_of_dofs);
+    qdd_target_prev.resize(kinematics.number_of_dofs);
+
+    l_next.resize(kinematics.number_of_cables);
+    l_target.resize(kinematics.number_of_cables);
+    ld.resize(kinematics.number_of_dofs);
+    for (int i = 0; i < kinematics.number_of_dofs; i++) {
+        ld[i].resize(kinematics.number_of_cables);
+        ld[i].setZero();
+    }
+
+    q_target.setZero();
+    qd_target.setZero();
+    qdd_target.setZero();
+
+    q_target_prev.setZero();
+    qd_target_prev.setZero();
+    qdd_target_prev.setZero();
+    l_next.setZero();
+    l_target.setZero();
+
+    controller_type.resize(kinematics.number_of_cables, CARDSflow::ControllerType::cable_length_controller);
+
+    /**
+     * Register ROS control
+     */
+    Kp_dl = new double(100.0);
+    Kd_dl = new double(0.0);
```

Comment:
add comments what it is, also rename to make it clear

Reply:
declare default values in the launch file

Reply:
Not resolved yet!